refactor: remove reconciliation methods, use runtime agent directly

enyst opened the pull request:
## Summary

This PR removes all reconciliation methods (`resolve_diff_from_deserialized`) and uses the provided Agent directly when restoring conversations. This is an alternative approach to issue #1451.

### What Was Happening on Main

On `main`, when restoring a conversation, `resolve_diff_from_deserialized` would:

**Override from runtime:**
- `agent_context` (skills, system_message_suffix, user_message_suffix, secrets)
- `llm` secrets (api_key, aws credentials, litellm_extra_body)
- `condenser.llm` secrets

**Restore from persistence (and require exact match with runtime):**
- `tools`
- `mcp_config`
- `filter_tools_regex`
- `system_prompt_filename`
- `security_policy_filename`
- `system_prompt_kwargs`
- `condenser` (except its llm secrets)
- `llm` config (model, temperature, etc.)

The final equality check meant users effectively couldn't change most Agent configuration between sessions.

### What This PR Does

**Removes reconciliation.** The provided Agent is used directly - subject to limitations that would otherwise not work at all, such as, it has to be the same Agent class, or the same tools.

Users are now free to change Agent configuration between sessions:
- `llm` (model, api_key, all settings)
- `mcp_config`
- `filter_tools_regex`
- `agent_context`
- `system_prompt_filename`
- `security_policy_filename`
- `system_prompt_kwargs`
- `condenser`

#### Limitations:
- `tools`
- Agent's class/type
- non-Agent `state` attributes like Confirmation Policy

### Execution Flow

#### New Conversation:
1. Create ConversationState with the provided Agent (Pydantic validation happens here)
2. Initialize EventLog for event storage
3. Save initial base state to persistence
4. Return the new state

#### Restored Conversation:
1. Load persisted base_state.json (to get conversation metadata)
2. Verify conversation ID matches
3. Create ConversationState with the provided Agent (Pydantic validation happens here)
4. Restore persisted conversation metadata (execution_status, confirmation_policy, etc.)
5. Attach EventLog to load persisted events
6. Save updated base state (with the provided Agent)
7. Return the resumed state

### Validation

Pydantic validation happens when creating instances (LLM, Agent, ConversationState) via the constructor.

### Note on Tools

From issue #1533 , tools already used in the conversation history are still available.

### Scope: LocalConversation Only

This PR only affects **LocalConversation**. 

For **RemoteConversation**, the server always creates the Agent from the persisted `meta.json` - the client's Agent is ignored when restoring. Making RemoteConversation support Agent changes would require:
1. Client sends new Agent config when attaching to existing conversation
2. Server accepts and uses the new Agent config instead of persisted one

This is out of scope for this PR but could be a follow-up.

Closes #1451

## Checklist

- [x] If the PR is changing/adding functionality, are there tests to reflect this?
- [x] If there is an example, have you run the example to make sure that it works?
- [x] If there are instructions on how to run the code, have you followed the instructions and made sure that it works?
- [ ] If the feature is significant enough to require documentation, is there a PR open on the OpenHands/docs repository with the same branch name?
- [ ] Is the github CI passing?



















































<!-- AGENT_SERVER_IMAGES_START -->
---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |


**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:521d39f-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-521d39f-python \
  ghcr.io/openhands/agent-server:521d39f-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:521d39f-golang-amd64
ghcr.io/openhands/agent-server:521d39f-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:521d39f-golang-arm64
ghcr.io/openhands/agent-server:521d39f-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:521d39f-java-amd64
ghcr.io/openhands/agent-server:521d39f-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:521d39f-java-arm64
ghcr.io/openhands/agent-server:521d39f-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:521d39f-python-amd64
ghcr.io/openhands/agent-server:521d39f-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-amd64
ghcr.io/openhands/agent-server:521d39f-python-arm64
ghcr.io/openhands/agent-server:521d39f-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-arm64
ghcr.io/openhands/agent-server:521d39f-golang
ghcr.io/openhands/agent-server:521d39f-java
ghcr.io/openhands/agent-server:521d39f-python
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `521d39f-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `521d39f-python-amd64`) are also available if needed
<!-- AGENT_SERVER_IMAGES_END -->

xingyaoww commented on openhands-sdk/openhands/sdk/agent/base.py:
```diff
@@ -300,72 +299,6 @@ def step(
-                f"Diff: {pretty_pydantic_diff(self, reconciled)}"
-            )
-        return reconciled
-
```
To me, it seems the whole goal of removing this "resolve_diff" is to increase the flexibility of dynamic configuration. One reason that i want to keep resolve diff function is that it give us control over what gets merged and what's not. IRRC, The real root cause that's causing us trouble was that we had a too strict default -- In this case, how about we simply relax the logic in `resolve_diff_from_deserialized` and allow overwritten most of the config, and log accordingly? 

In practice, it will behave the same as this PR, just with fewer code changes.

It will also make it easier to implement https://github.com/OpenHands/software-agent-sdk/issues/1533, since one only need to add a few lines in this consolidation function to tighten up the loading of particular attributes?

WDYT?

xingyaoww commented on openhands-sdk/openhands/sdk/conversation/state.py:
```diff
@@ -191,26 +216,42 @@ def create(
-            resolved = agent.resolve_diff_from_deserialized(state.agent)
+            # Limit to the tools that match - they may have been used in conversation history.
+            # All other config (LLM, agent_context, condenser, etc.) can change freely.
+            agent.load(persisted_state.agent)
```
Shall we get the return value of this function?

xingyaoww commented on openhands-sdk/openhands/sdk/conversation/state.py:
```diff
@@ -191,26 +216,42 @@ def create(
+                blocked_actions=persisted_state.blocked_actions,
+                blocked_messages=persisted_state.blocked_messages,
+                secret_registry=persisted_state.secret_registry,
+            )
```
Instead of creating this from scratch, can we just explicitly overwrite the few keys for `persisted_states`, in case we add more fields in the future but forgot to add a new key here

enyst commented on openhands-sdk/openhands/sdk/agent/base.py:
```diff
@@ -300,49 +299,36 @@ def step(
-        Return a new AgentBase instance equivalent to `persisted` but with
-        explicitly whitelisted fields (e.g. api_key) taken from `self`.
+    def load(self, persisted: "AgentBase") -> "AgentBase":
+        """Validate that this agent can resume a conversation from persisted state.
```
```suggestion
        """Verify that we can resume this agent from persisted state.
```

enyst commented on openhands-sdk/openhands/sdk/conversation/state.py:
```diff
@@ -193,22 +221,27 @@ def create(
             state = cls.model_validate(json.loads(base_text))
 
-            # Enforce conversation id match
             if state.id != id:
```
```suggestion
            # Restore the conversation with the same id
            if state.id != id:
```

enyst commented:
@openhands Understand the goal that this PR aims to achieve. We have two tests files, modified in this PR, which are no longer named appropriately since we removed reconciliation and replace it with "verification". Rename test_agent_reconciliation.py to test_agent_loading.py (or verification if you prefer, but maybe loading will do, wdyt?)

Review the tests in tests/sdk/llm/test_llm_reconciliation.py and if they're not duplicated in other tests/sdk llm-related files and they're still useful (are they?), then move them in a test llm file.

enyst commented:
@openhands Understand the goal of this PR and its diff. We need to merge main into it, and fix the conflicts. Find the recent commit that introduced the conflicts (it's very recent, last five or so) and understand its goal too, so we can keep that behavior as well as this one.

Then do it, merge and fix.

enyst commented on tests/integration/tests/t10_restore_conversation.py:
```diff
@@ -0,0 +1,178 @@
+                reason="confirmation_policy was not restored from persistence",
+            )
+
+        # 2) Runtime agent/LLM should be used (no reconciliation).
```
```suggestion
        # 2) Runtime agent/LLM should be used.
```

enyst commented on openhands-sdk/openhands/sdk/agent/base.py:
```diff
@@ -366,43 +353,23 @@ def resolve_diff_from_deserialized(
             }
 
-            # Only require tools that were actually used in history
             missing_used_tools = used_tools - runtime_names
```
```suggestion
            # Only require tools that were actually used in history
            missing_used_tools = used_tools - runtime_names
```

enyst commented on tests/cross/test_agent_loading.py:
```diff
@@ -106,6 +106,7 @@ def test_conversation_restarted_with_changed_working_directory(tmp_path_factory)
 
 
+# # Tests for agent tools restriction and LLM flexibility
 # Tests from test_local_conversation_tools_integration.py
```
```suggestion
# Tests for agent tools restriction and LLM flexibility
```

enyst commented on tests/cross/test_agent_loading.py:
```diff
@@ -458,6 +452,75 @@ def test_conversation_persistence_lifecycle(mock_completion):
+    - Some conversation *state* settings are persisted and should not be overridden
+      on resume (e.g., confirmation_policy, execution_status).
+    - Agent/LLM settings should come from the runtime-provided Agent on resume
+      (no reconciliation).
```
```suggestion
```

enyst reviewed (commented):
@xingyaoww We made here both `cross` test and `integration test`. Both ran and worked repeatedly in agent's workspace.

I'm not sure about the integration test TBH. 🤔 Verifying the correct behavior on conversation restore doesn't really depend on LLM; so running it repeatedly for 6 LLMs seems a bit ... not so useful perhaps? What it needs is at most 2 in use during the test itself.

**Options:**

**1. Two real LLMs**
if we want real LLM calls, and with 2 LLMs, the agent suggested something like:

  - Add a new job in .github/workflows/integration-runner.yml, something like `run-restore-test-once`:
    - runs only when label is integration-test (and maybe schedule / dispatch)
    - sets env vars:
        - LLM_CONFIG_1 = first model config from matrix (e.g. Sonnet)
        - LLM_CONFIG_2 = second model config from matrix (e.g. GPT-5.1-codex-max)
    - runs:
        - `uv run python tests/integration/run_infer.py --eval-ids t10_restore_conversation --llm-config "$LLM_CONFIG_1" ...`
        - and passes the “second model” to the test via an env var (e.g. `RESTORE_LLM_CONFIG_2`), or via a dedicated CLI arg if we extend `run_infer.py`.
  - In t10, stop hardcoding model names; instead:
    - use self.llm for llm1 (this is provided by --llm-config)
    - use the second config from env/arg for llm2 (the “other model”)
    - keep using the same LLM_API_KEY / LLM_BASE_URL from CI secrets.

**2. One real LLM**
 - alternatively, if we want with a single LLM, but still actually run `send_message` for real, we can do only the first part above, and avoid the roundabout with the `RESTORE_LLM_CONFIG_2` part

**3. Not run real LLMs**
- run a `cross` test verifying the restore conversation behavior. (the right attributes, some from persisted, some from runtime)

WDYT?
